### PR TITLE
Fetch VSphereCluster by labels in non-paravirtual mode for vsphere csi/cpi config controller

### LINKED ISSUE #3223: vsphere csi/cpi config controller is not able to retrieve the correct VSphereCluster CR in non-paravirtual mode
**Bug description**
Management cluster creation for clusterclass based vsphere cluster results in package install errors for both vsphere csi and cpi:
```
E0826 21:17:19.331272       1 controller.go:317] controller/cluster "msg"="Reconciler error" "error"="failed to get remote cluster client: error creating dynamic rest mapper for remote cluster \"tkg-system/tkg-mc-vsphere\": Get \"https://10.184.90.139:6443/api?timeout=10s\": dial tcp 10.184.90.139:6443: i/o timeout" "name"="tkg-mc-vsphere" "namespace"="tkg-system" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="Cluster" 
...
E0826 21:17:21.857637       1 vspherecsiconfig_controller.go:250] VSphereCSIConfig "msg"="Error while mapping VSphereCSIConfig to data values" "error"="VSphereCluster tkg-system/tkg-mc-vsphere not found" "VSphereCSIConfig"={"Namespace":"tkg-system","Name":"tkg-mc-vsphere-vsphere-csi-package"} 
...
E0826 21:45:43.630401       1 vspherecpiconfig_controller.go:171] VSphereCPIConfig "msg"="Error while mapping VSphereCPIConfig to data values" "error"="VSphereCluster tkg-system/tkg-mc-vsphere not found" "VSphereCPIConfig"={"Namespace":"tkg-system","Name":"tkg-mc-vsphere"} 
```

**Affected product area (please put an X in all that apply)**

- ( ) APIs
- (x) Addons
- ( ) CLI
- ( ) Docs
- ( ) IAM
- ( ) Installation
- ( ) Plugin
- ( ) Security
- ( ) Test and Release
- ( ) User Experience
- ( ) Developer Experience

**Expected behavior**
VSphereCluster is mapped correctly

**Steps to reproduce the bug**
Create cc based management cluster on vsphere and observe that CSI/CPI packages are not installed successfully.

**Version** (include the SHA if the version is not obvious)

**Environment where the bug was observed (cloud, OS, etc)**

**Relevant Debug Output (Logs, manifests, etc)**

### PR DESCRIPTION
**Pull request by haoranleo**
### What this PR does / why we need it
- fetches `VSphereCluster` CR by labels in non-paravirtual mode for vsphere csi/cpi config controller

### Which issue(s) this PR fixes

Fixes #3223 

### Describe testing done for PR
Manual testing In progress.

### Release note

```release-note
Fetches VSphereCluster CR by labels in non-paravirtual mode for vsphere csi/cpi config controller
```

### CODE REVIEW

File: addons/controllers/utils/utils.go [hunk 1]
```diff
@@ -11,28 +11,56 @@ import (
...
-	if err := clt.Get(ctx, types.NamespacedName{
-		Namespace: cluster.Namespace,
-		Name:      cluster.Name,
-	}, vsphereCluster); err != nil {
-		if apierrors.IsNotFound(err) {
-			return nil, errors.Errorf("VSphereCluster %s/%s not found", cluster.Namespace, cluster.Name)
-		}
-		return nil, errors.Errorf("VSphereCluster %s/%s could not be fetched, error %v", cluster.Namespace, cluster.Name, err)
+// GetVSphereClusterParavirtual gets the VSphereCluster CR for the cluster object in paravirtual mode
+func GetVSphereClusterParavirtual(ctx context.Context, clt client.Client, cluster *clusterapiv1beta1.Cluster) (*capvvmwarev1beta1.VSphereCluster, error) {
```
Comment (vijaykatam) [inline addons/controllers/utils/utils.go:27]: lets add some unit tests for these 2 functions

**Review by vijaykatam — COMMENTED** [review]
Add unit tests and this is good to merge.

File: addons/controllers/utils/utils.go (hunk 1 above)
Comment (HL-EverGreen) [reply addons/controllers/utils/utils.go:27]: Ah sorry Vijay I missed your comment. Will add unit tests. Thanks for your quick review!

Comment (HL-EverGreen) [reply addons/controllers/utils/utils.go:27]: Hi Vijay, current tests are modified to cover the scenario of non-paravirtual mode for vsphere csi/cpi config controller.

File: addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml [hunk 2]
```diff
@@ -61,6 +61,8 @@ data:
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
 kind: VSphereCluster
 metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: test-cluster-cpi
   name: test-cluster-cpi
```
Comment (vijaykatam) [inline addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml:66]: Could we update the cluster to mimic how capv adds a suffix?

File: addons/controllers/testdata/test-vsphere-csi-non-paravirtual-minimal.yaml [hunk 3]
```diff
@@ -63,6 +63,8 @@ data:
 apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
 kind: VSphereCluster
 metadata:
+  labels:
+    cluster.x-k8s.io/cluster-name: test-cluster-csi-minimal
   name: test-cluster-csi-minimal
```
Comment (vijaykatam) [inline addons/controllers/testdata/test-vsphere-csi-non-paravirtual-minimal.yaml:68]: (same comment): Could we update the name of VSphereCluster to mimic how capv adds a suffix?

File: addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml (hunk 2 above)
Comment (HL-EverGreen) [reply addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml]: Thanks for pointing that out. `VSphereCluster` names are updated.

File: addons/controllers/cpi/vspherecpiconfig_utils.go [hunk 4]
```diff
@@ -377,22 +378,6 @@ func (r *VSphereCPIConfigReconciler) getOwnerCluster(ctx context.Context, cpiCon
...
-		if apierrors.IsNotFound(err) {
-			return nil, errors.Errorf("VSphereCluster %s/%s not found", cluster.Namespace, cluster.Name)
-		}
-		return nil, errors.Errorf("VSphereCluster %s/%s could not be fetched, error %v", cluster.Namespace, cluster.Name, err)
-	}
-	return vsphereCluster, nil
-}
-
 // getSecret gets the secret object given its name and namespace
 func (r *VSphereCPIConfigReconciler) getSecret(ctx context.Context, namespace, name string) (*v1.Secret, error) {
```
Comment (vijaykatam) [inline addons/controllers/cpi/vspherecpiconfig_utils.go:382]: This line for fetching VsphereMachineTemplate needs the same fix. https://github.com/vmware-tanzu/tanzu-framework/blob/86a0e3d6dbd2b6ef1fccc006ae1d93f55e5f5cea/addons/controllers/cpi/vspherecpiconfig_utils.go#L62

File: addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml (hunk 2 above)
Comment (DanielXiao) [reply addons/controllers/testdata/test-vsphere-cpi-non-paravirtual.yaml]: Why don't we get these information from Cluster variables? This is an example of variable vcenter 
```
- name: vcenter
      value:
        cloneMode: fullClone
        datacenter: /dc0
        datastore: /dc0/datastore/sharedVmfs-0
        folder: /dc0/vm
        network: /dc0/network/VM Network
        resourcePool: /dc0/host/cluster0/Resources
        server: 10.187.96.70
        storagePolicyID: ""
        template: /dc0/vm/photon-3-kube-v1.23.8+vmware.2-tkg.2-81d1a7892ad39f017fbaf59f9907cbe7
        tlsThumbprint: 1E:BB:09:45:FD:02:10:17:FD:55:8D:D1:05:8B:3F:A8:91:DD:29:0E
```

File: addons/controllers/cpi/vspherecpiconfig_utils.go (hunk 4 above)
Comment (DanielXiao) [reply addons/controllers/cpi/vspherecpiconfig_utils.go:382]: The same as above, we can get datacenter from the following variable:
```
- name: vcenter
      value:
        cloneMode: fullClone
        datacenter: /dc0
        datastore: /dc0/datastore/sharedVmfs-0
        folder: /dc0/vm
        network: /dc0/network/VM Network
        resourcePool: /dc0/host/cluster0/Resources
        server: 10.187.96.70
        storagePolicyID: ""
        template: /dc0/vm/photon-3-kube-v1.23.8+vmware.2-tkg.2-81d1a7892ad39f017fbaf59f9907cbe7
        tlsThumbprint: 1E:BB:09:45:FD:02:10:17:FD:55:8D:D1:05:8B:3F:A8:91:DD:29:0E
```

**Review by DanielXiao — COMMENTED** [review]
There is no Cluster variable `VSPHERE_NETWORK`, you can get publicNetwork from variable `vcenter.network`:
```
    - name: vcenter
      value:
        cloneMode: fullClone
        datacenter: /dc0
        datastore: /dc0/datastore/sharedVmfs-0
        folder: /dc0/vm
        network: /dc0/network/VM Network
        resourcePool: /dc0/host/cluster0/Resources
        server: 10.187.96.70
        storagePolicyID: ""
        template: /dc0/vm/photon-3-kube-v1.23.8+vmware.2-tkg.2-81d1a7892ad39f017fbaf59f9907cbe7
        tlsThumbprint: 1E:BB:09:45:FD:02:10:17:FD:55:8D:D1:05:8B:3F:A8:91:DD:29:0E
```
CSI package can not be installed if `publicNetwork` is not set correctly. Likewise we need to change other legacy-style variables to classy Cluster variables.

**Comment by vijaykatam** [comment]
> There is no Cluster variable `VSPHERE_NETWORK`, you can get publicNetwork from variable `vcenter.network`:
> 
> ```
>     - name: vcenter
>       value:
>         cloneMode: fullClone
>         datacenter: /dc0
>         datastore: /dc0/datastore/sharedVmfs-0
>         folder: /dc0/vm
>         network: /dc0/network/VM Network
>         resourcePool: /dc0/host/cluster0/Resources
>         server: 10.187.96.70
>         storagePolicyID: ""
>         template: /dc0/vm/photon-3-kube-v1.23.8+vmware.2-tkg.2-81d1a7892ad39f017fbaf59f9907cbe7
>         tlsThumbprint: 1E:BB:09:45:FD:02:10:17:FD:55:8D:D1:05:8B:3F:A8:91:DD:29:0E
> ```
> 
> CSI package can not be installed if `publicNetwork` is not set correctly. Likewise we need to change other legacy-style variables to classy Cluster variables.

Why do you think we need to use Cluster variables as opposed to attributes from linked Vsphere* objects? Is there info in Cluster variables that is not available in Vsphere* objects?

**Comment by HL-EverGreen** [comment]
Close this PR since the commit is included in https://github.com/vmware-tanzu/tanzu-framework/pull/3271

**Pull request closed without merging**